internal/ethapi/api.go

```diff
@@ -1202,35 +1188,56 @@ func DoEstimateGas(ctx context.Context, b Backend, args TransactionArgs, blockNr
 		log.Warn("Caller gas above allowance, capping", "requested", hi, "cap", gasCap)
 		hi = gasCap
 	}
-	cap = hi
 
-	// Create a helper to check if a gas allowance results in an executable transaction
-	executable := func(gas uint64, state *state.StateDB, header *types.Header) (bool, *core.ExecutionResult, error) {
+	// Create a helper that executes the transaction under a given gas limit and returns true
+	// if the transaction fails for a reason that might be related to not enough gas. A non-nil
+	// error means execution failed due to reasons unrelated to the gas limit.
+	executable := func(gas uint64, state *state.StateDB) (bool, *core.ExecutionResult, error, uint64) {
 		args.Gas = (*hexutil.Uint64)(&gas)
 
-		result, err := doCall(ctx, b, args, state, header, nil, nil, 0, gasCap)
+		result, gasUsed, err := doCall(ctx, b, args, state, header, nil, nil, 0, gasCap)
 		if err != nil {
 			if errors.Is(err, core.ErrIntrinsicGas) {
-				return true, nil, nil // Special case, raise gas limit
+				return true, nil, nil, 0 // Special case, raise gas limit
 			}
-			return true, nil, err // Bail out
+			return true, nil, err, 0 // Bail out
 		}
-		return result.Failed(), result, nil
+		return result.Failed(), result, nil, gasUsed
 	}
-	state, header, err := b.StateAndHeaderByNumberOrHash(ctx, blockNrOrHash)
-	if state == nil || err != nil {
+
+	// We first execute the transaction at the highest allowable gas limit, since if this fails we
+	// can return error immediately.
+	failed, result, err, gasUsed := executable(hi, st.Copy())
+	if err != nil {
 		return 0, err
 	}
-	// Execute the binary search and hone in on an executable gas limit
+	if failed {
+		if result != nil && result.Err != vm.ErrOutOfGas {
+			if len(result.Revert()) > 0 {
+				return 0, newRevertError(result)
+			}
+			return 0, result.Err
+		}
+		return 0, fmt.Errorf("gas required exceeds allowance (%d)", hi)
+	}
+	// For almost any transaction, the gas consumed by the unconstrained execution above
+	// lower-bounds the gas limit required for it to succeed. One exception is those txs that
+	// explicitly check gas remaining in order to successfully execute within a given limit, but we
+	// probably don't want to return a lowest possible gas limit for these cases anyway.
+	lo = gasUsed - 1
+
+	// Binary search for the smallest gas limit that allows the tx to execute successfully.
 	for lo+1 < hi {
-		s := state.Copy()
 		mid := (hi + lo) / 2
-		failed, _, err := executable(mid, s, header)
-
-		// If the error is not nil(consensus error), it means the provided message
-		// call or transaction will never be accepted no matter how much gas it is
-		// assigned. Return the error directly, don't struggle any more.
+		if mid > lo*2 {
```

Comment:
There is likely room for further average-case improvement by tweaking (lowering) the constant multiplier here, but this would worsen the worst-case and also introduces new boundary conditions that would need to be checked.

Reply:
This is more of an acceptance that there are a much larger number of transactions on the low size, 21k - 500k gas?

Reply:
bit of both I guess: most txs don't need much higher gas limit than their gas used, and most txs don't require near the full block limit of gas. I guess this trick takes advantage of both facts. (FWIW EstimateGas doesn't seem to be called for simple 21k gas transfers by wallets like MetaMask so I excluded those from the analysis)